### ISSUE TITLE
Problems encountered processing cataloge keywords. Timed out after 30 seconds

### ISSUE THREAD

**Issue by nyomanb** [mcrosson/lr_plugin_computer_vision_tagging#19]
Hello,

thanks for developing this plugin. It looks really nice, but unfortunately it has stopped working for me after uploading a few test images with Clarifai. 

After exporting only the error message "[An internal error has occured.] Problems encountered processing cataloge keywords. Timed out after 30 seconds" appears. Timed out after 30 seconds" appears. The log file (saved under C:\Users\ben\Documents) is not really helpful: 

> 10/20/2017 21:10:48 INFO	Debug log enabled 
10/20/2017 21:10:54 TRACE	exportServiceProvider.sectionsForTopOfDialog(vf, propertyTable) 
10/20/2017 21:10:56 TRACE	exportServiceProvider.updateExportSettings(exportSettings) 
10/20/2017 21:10:56 TRACE	exportServiceProvider.processRenderedPhotos(functionContext, exportContext) 
10/20/2017 21:10:56 TRACE	LUTILS.waitForGlobal(globalName, timeout) 
10/20/2017 21:11:26 TRACE	Global _G.AllKeys non-existent after waiting 30 seconds 

I have saved about 31,500 keywords in Lightroom. Maybe a few too many for this plugin? (But disabling "Select/Bold existing keywords/tags" did not help). Any ideas?

Cheers,
Ben

**Comment by LowellMontgomery** [comment]
Hi Ben,

I wrote the code that iterates the tree so that this plugin could work for keyword sets that had any kind of hierarchy (otherwise it simply selected—or created new terms—at the base level, which would be frustrating if you had an appropriate term in your tree (if you added the term at the base level, then you'd be faced with "merging" the terms (selecting the appropriate term from the hierarchy, e.g. "Color > Gold", then deselecting the new term, "Gold", at the base of the keyword tree). Manually merging the terms was such a tedious chore, I jumped into the code to figure out a solution... which appears to be what you're running into. Maybe my "solution" doesn't scale well enough.

That (31,500) is a LOT of keywords that Lightroom is trying to parse before it starts its work. If your keywords are not hierarchical, the work it's doing is unnecessary. Otherwise, we probably need to provide a setting for that timeout (to allow the pre-processing of your keyword tree to take longer than 30 seconds). Assuming your keywords ARE hierarchical, and that some branches of your keyword tree include a lot of terms that an automated system like Clarifai would never return, you can list terms that plugin will not iterate. For instance, I have long lists of bird and plant species, people (by name), place names (within hierarchies of other location keywords). All of these things are far too specific to normally be returned by Clarifai, so I list them in the "Ignore keywords branches" field (e.g. _IMAGE_LOCATIONS, BIRD_TYPES, dog breeds, Mammals, MY_PEOPLE, Plants-by-Taxonomy). Note that the ignored terms should be unique in your keyword tree, but they can appear at any level in the hierarchy (they simply terminate further requests to the Lightroom API to find "child" keywords for any of these terms or their children). **So skip as much of your keyword tree as you can**, especially if there are big chunks of your keyword tree with terms you'd never get back from the system (or which, if they did match, would only be out-of-context). For instance, I skip all sections of the "LOCATIONS", partly also because the plugin was otherwise often providing checkboxes for "downtown", etc. (Since "downtown" was a child keyword within many city/location keywords, existing "downtown" keywords would normally be out-of-context for any new photo.)

Also, do bear in mind that this process is not just about the "Select/Bold existing keywords/tags"; indeed, that doesn't contribute to the timeout at all. A checkbox still has to either correspond to an existing keyword (there may be more than one checkbox for the same term if two keywords share a name, e.g. "Metal > Gold" and "Color > Gold") or correspond to the action of creating a new keyword (at the root of your hierarchy or as a "child" of any keyword you've set to be the parent of new keywords, i.e., by using the contextual menu setting "Put New Keywords Inside this Keyword"). I definitely do recommend creating a "New Keywords" parent, assuming you have a hierarchical keyword list, so that your "root level" doesn't get polluted by all the new terms. Indeed, even if your keyword list is otherwise "flat", I'd recommend putting new keywords (added by a process like this) under a parent term that would allow you to take some time to potentially merge or re-organize your keywords. For instance, to better work with the Clarifai system, I renamed some of my keywords to match the terms they returned (and added my original keyword term as a "synonym" that would be automatically included on export).

Anyway, the logic related to the timeout error is in the following block of code, within the file, `DialogTagging.lua`. You can change the `local timeout = 30;` to any value (maybe 60?) and see if that is enough time for the process (of iterating all of your keywords) to complete. I'm hoping that increasing the timeout should solve the problem, but it's also possible that so many keywords results in some other issue that causes the process of building up the keyword tree to fail. Assuming you can get this working by changing the timeout, then we definitely want to provide a setting for that (in the Plugin Manager panel), so that users don't need to dig around in the code. But since this process needs to take place each time the plugin is run, you'd want to submit larger batches of images at a time (so that the process of iterating your keyword tree hopefully doesn't take much longer than the time required to export images to Clarifai and get results back).

```lua
    -- Before we trim down our lookup table for keywords and keyword paths, we should
    -- be sure the process of populating our global variables for these has completed.
    local timeout = 30;
    local timeWaited = LUTILS.waitForGlobal('AllKeys', timeout);
    
    if timeWaited and timeWaited < timeout then
      KmnUtils.log(KmnUtils.LogTrace, 'Global _G.AllKeys ready for use after ' .. timeWaited .. ' seconds');
    elseif timeWaited == false then
      KmnUtils.log(KmnUtils.LogTrace, 'Global _G.AllKeys non-existent after waiting ' .. timeout .. ' seconds');
      LrDialogs.showError('Problems encountered processing catalog keywords. Timed out after ' .. timeout .. ' seconds');
      return
    end
```

I'm hoping I've given you enough information you may be able to get things working now, but please let me know if that's not the case or if you run into anything "unexpected" (which, in your situation, with as many keywords as you have, might actually be somewhat "expected").

**Comment by LowellMontgomery** [comment]
tl;dr? 😉

In short:
- If your tree is hierarchical, skip sections that have terms that Clarifai would never return. (Use the "Ignore keywords branches" configuration field.)
- If the script still times out, increase the "timeout value" (e.g. try changing `local timeout = 30;` to  `local timeout = 60;` or even 300... and see if that makes a difference.
- (Oh, and, I forgot to mention... ) Use Lightroom's Plugin Manager panel to **"Reload Plug-in" between any changes you make to the plugin code** (otherwise Lightroom will not recognize the code changes and will continue to run whatever it loaded on start-up).

It's possible that the timeout may be symptomatic of exceeding other internal limits (which could involved the version of Lightroom, your RAM/processors/OS, etc). If that seems to be the case, we can easily improve the debug info to gain a better understanding and maybe figure out a workaround.

**Comment by nyomanb** [comment]
Thanks for your detailed answer! For testing purposes I've added all top level keywords (I am using a hierarchical keyword list) to the "Ignore keywords branches" configuration field. This did not help. However, changing the timeout value to 120 seconds did help! :-) Lightroom still needs a long time to begin processing the images after starting the export -but this is okay for me!

This issue is probably related to my use case - using a large keyword list with an older computer (Intel Q6600 CPU; 5 GB RAM; Win10, Lightroom 6). As the fix is easy to implement and working, there's probably no need to add an extra timeout field to the plugin settings.

All in all, thank you for developing this plugin and especially for your great support!

Greetings from Germany,
Ben

**Comment by mcrosson** [comment]
@nyomanb I wouldn't be opposed to updating the plugin for a timeout field. Especially since it's going to add some useful value to those who might be running older hardware.

**Comment by LowellMontgomery** [comment]
@mcrosson Hey, Mike. What I'm thinking might be useful is something I could add to my code, also, so that if the process running in `KwUtils.lua` hasn't completed on time, we can get more insight to the status of that process. When we DO end in a timeout, the user should have better clarity about WHAT process was taking too long. I'll go ahead and make some improvements and send you a pull request for those. But I probably won't get to that right away since I may also take the time to do another refactoring of my code. My improvements will mainly be to the "helper functions" (in LUTILS.lua and KwUtils.lua), but my PR will include some changes in the tagging dialog (to implement the improvements in my code, etc).

This will add some value to people with larger catalogs as well as those running on older hardware.

@mcrosson Anyway, if you want to start by providing a field for that keyword-parsing timeout, I guess that should accept up to 4-5 digits, have a default value of 30, and some help text explaining it, and of course use that field's value to init the timeout variable. I'll assume those changes are being made "on your end". But please leave this ticket open till I've given you a PR that will improve the UX for this "edge case".
